### PR TITLE
Dimensionality Reduction Updates

### PR DESCRIPTION
**Pull request by akshayparopkari**
## Description

- In LDA:
  - Added checks for invalid use of 3D figures.
  - Removed sample annotation for 3D figures, now it is only available for 2D images.
  - Annotations now support 1D figures as well.
  - Parameter `—otu_table` is now optional.
  - Updated Pandas method to extract input values for calculations from dataframe.
  - Updated documentation.

- In PCoA:
  - Added edge colors to scatter plots.

- All figures were tested on matplotlib version 2.0.2.

## Motivation and Context

- Compatibility with matplotlib >=v2.0
- Better checks for discriminant analysis, providing users useful feedback for ease of use.
- #47 - User-friendly 3D LDA visualization process

## How Has This Been Tested?

- All tests using command `python -m unittest discover -s phylotoast/test/ -vf` ran successfully.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read  he **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: bin/LDA.py [hunk 1]
```diff
@@ -51,6 +50,14 @@ def plot_LDA(X_lda, y_lda, class_colors, exp_var, style, fig_size, label_pad,
     cats = class_colors.keys()
     fig = plt.figure(figsize=fig_size)
     if dim == 3:
+        try:
+            assert X_lda.shape[1] >= 3
+        except AssertionError:
+            sys.exit("\nLinear Discriminant Analysis requires atleast 4 groups of samples"
```
Comment (smdabdoub) [inline bin/LDA.py]: Needs a space in "atleast"

File: bin/LDA.py [hunk 2]
```diff
@@ -51,6 +50,14 @@ def plot_LDA(X_lda, y_lda, class_colors, exp_var, style, fig_size, label_pad,
...
+                     " to create a 3D figure. Please update group information or use the "
+                     "default 2D view of the results.\n")
+        if sids is not None:
+            print("\nPoint annotations are available only for 2D figure.\n")
```
Comment (smdabdoub) [inline bin/LDA.py]: figures

File: bin/LDA.py [hunk 3]
```diff
@@ -123,7 +127,7 @@ def run_LDA(df):
     transformed data, scalings and explained variance by discriminants.
     """
     # Prep variables for sklearn LDA
-    X = df[range(1, df.shape[1])].values     # input data matrix
+    X = df.iloc[:, 1:df.shape[1]].values     # input data matrix
```
Comment (smdabdoub) [inline bin/LDA.py]: Nice change.

File: bin/LDA.py [hunk 4]
```diff
@@ -143,48 +147,44 @@ def handle_program_options():
...
-                        help="A column name in the mapping file containing\
-                              hexadecimal (#FF0000) color values that will\
-                              be used to color the groups. Each sample ID must\
-                              have a color entry.")
-    parser.add_argument("-dm", "--dist_matrix_file",
-                        help="Input distance matrix file.")
+                        help="A column name in the mapping file containing hexadecimal "
+                             "(#FF0000) color values that will be used to color the "
+                             "groups. Each sample ID must have a color entry.")
+    parser.add_argument("-i", "--otu_table", help="Input biom file format OTU table.")
```
Comment (smdabdoub) [inline bin/LDA.py]: Now that otu_table is no longer the "default" expected input, should we change from -i to something like -ot to match the -dm argument?

File: bin/LDA.py [hunk 5]
```diff
@@ -207,26 +207,23 @@ def main():
...
-        class_colors = {c:color_cycle.next() for c in categories}
+        class_colors = {c: color_cycle.next() for c in categories}
     else:
         class_colors = util.color_mapping(imap, header, args.group_by, args.colors)
 
     if args.dist_matrix_file:
         try:
-            with open(args.dist_matrix_file):
-                pass
+            uf_data = pd.read_csv(args.dist_matrix_file, sep="\t", index_col=0)
```
Comment (smdabdoub) [inline bin/LDA.py]: if uf here stands for unifrac, it might be better to name it dm_data

File: docs/scripts/LDA.txt [hunk 6]
```diff
@@ -6,15 +6,11 @@ This script calculates and returns LDA plots based on normalized relative abunda
 
 .. code-block:: bash
 
-        usage: LDA.py [-h] -i OTU_TABLE -m MAP_FP -g GROUP_BY -c COLOR_BY [-dm DIST_MATRIX_FILE] [--save_lda_input SAVE_LDA_INPUT] [--plot_title PLOT_TITLE] [-o OUT_FP] [-d {2,3}] [--z_angles Z_ANGLES Z_ANGLES] [--figsize FIGSIZE FIGSIZE] [--font_size FONT_SIZE] [--label_padding LABEL_PADDING] [--ggplot2_style]
+        usage: LDA.py [-h] -m MAP_FP -g GROUP_BY [-c COLORS] [-i OTU_TABLE] [-dm DIST_MATRIX_FILE] [--save_lda_input SAVE_LDA_INPUT] [--plot_title PLOT_TITLE] [-o OUT_FP] [-d {2,3}] [--z_angles Z_ANGLES Z_ANGLES] [--figsize FIGSIZE FIGSIZE] [--font_size FONT_SIZE] [--label_padding LABEL_PADDING] [--annotate_points] [--ggplot2_style]
```
Comment (smdabdoub) [inline docs/scripts/LDA.txt]: Same note as above with regard to arg name for otu_table

**Review by smdabdoub — CHANGES_REQUESTED** [review]
A few minor changes and some suggestions.

File: bin/LDA.py (hunk 1 above)
Comment (akshayparopkari) [inline bin/LDA.py]: done

File: bin/LDA.py (hunk 2 above)
Comment (akshayparopkari) [inline bin/LDA.py]: done

File: bin/LDA.py (hunk 3 above)
Comment (akshayparopkari) [inline bin/LDA.py]: Pandas update broke the earlier method of getting 2D array.

File: bin/LDA.py (hunk 4 above)
Comment (akshayparopkari) [inline bin/LDA.py]: yes, I updated the argument list.

File: bin/LDA.py (hunk 5 above)
Comment (akshayparopkari) [inline bin/LDA.py]: Updated

File: docs/scripts/LDA.txt (hunk 6 above)
Comment (akshayparopkari) [inline docs/scripts/LDA.txt]: Updated

**Comment by akshayparopkari** [comment]
Review changes incorporated. Ready for a second pass.

**Review by smdabdoub — APPROVED** [review]

**Pull request merged**